src/vscode-dts/vscode.proposed.tunnelFactory.d.ts

```diff
@@ -0,0 +1,44 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+declare module 'vscode' {
+	/**
+	 * Used as part of the ResolverResult if the extension has any candidate,
+	 * published, or forwarded ports.
+	 */
+	export interface TunnelInformation {
+		/**
+		 * Tunnels that are detected by the extension. The remotePort is used for display purposes.
+		 * The localAddress should be the complete local address (ex. localhost:1234) for connecting to the port. Tunnels provided through
+		 * environmentTunnels are read-only from the forwarded ports UI.
+		 */
+		environmentTunnels?: TunnelDescription[];
```

Comment:
I didn't implement this. It is copied over from `resolvers` and let's the tunnel provider define some readonly ports at the time of registration. Is this useful?